### PR TITLE
[DO-2471]Made Person model

### PR DESCRIPTION
**Pull request by IegorSoboliev**

### CODE REVIEW

**Comment by IegorSoboliev** [comment]
Please, look and after all improvements, I will add the migration)

**Review by roman-tiukh — CHANGES_REQUESTED** [review]

**Review by EmmaSamoilova — APPROVED** [review]

File: person/models.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+                               CitizenshipStatus
+                               )
+from location_register.models.address_models import Country
+
+
+class Person(DataOceanModel):
+    first_name = models.CharField(
+        _('first name'),
+        max_length=50,
+        validators=[RegexValidator(regex=r'[A-Za-z]')],
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:18]: added

[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:18]: and made custom validator

File: person/models.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+from data_ocean.models import (DOBAccuracy,
+                               Gender,
+                               CitizenshipStatus
+                               )
+from location_register.models.address_models import Country
+
+
+class Person(DataOceanModel):
+    first_name = models.CharField(
+        _('first name'),
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:16]: deleted)

File: person/models.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        help_text='last name of the person in the language of the country of birth or residence',
+    )
+    full_name_original = models.CharField(
+        _('full name original'),
+        blank=True,
+        default='',
+        max_length=150,
+        help_text='full name of the person in the language of the country of birth or residence',
+    )
+    full_name_options = ArrayField(
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:69]: changed

File: person/models.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            default='',
+            validators=[RegexValidator(regex=r'[A-Za-z]')],
+            max_length=150
+        ),
+        verbose_name=_('full name options'),
+        default=list,
+        help_text='options for spelling the full name of the person',
+    )
+    date_of_birth = models.DateField(
+        _('date of birth  of the person'),
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:81]: thank you, fixed

File: person/models.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        choices=DOBAccuracy.choices,
+        default=DOBAccuracy.EXACT,
+        help_text=_('date of birth accuracy'),
+
+    )
+    gender = models.PositiveSmallIntegerField(
+        _('gender of the person'),
+        choices=Gender.choices,
+        help_text=_('gender of the person')
+    )
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:98]: added as default

File: person/models.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    is_dead = models.BooleanField(
+        _('is dead'),
+        default=False,
+        help_text=_('is person dead')
+    )
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:103]: added as default

File: person/models.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    citizenship = models.ManyToManyField(
+        Country,
+        related_name='citizens',
+        verbose_name=_('countries of citizenship'),
+        help_text=_('countries of citizenship of the person')
+    )
+    residence = models.ForeignKey(
+        Country,
+        on_delete=models.CASCADE,
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:117]: changed

File: person/models.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    taxpayer_number = models.CharField(
+        _('taxpayer number'),
+        max_length=20,
+        blank=True,
+        default='',
+        help_text=_('taxpayer number of the person'),
+    )
+    is_main = models.BooleanField(
+        _('is main'),
+        default=True,
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:137]: changed

File: person/models.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        related_name='duplicates',
+        null=True,
+        blank=True,
+        default=None,
+        verbose_name=_('duplicate of the person'),
+        help_text=_('duplicate of the record of the person'),
+    )
+
+    def __str__(self):
+        return f'{self.full_name} from {self.residence}'
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:152]: changed

File: person/models.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    )
+    duplicate_of = models.ForeignKey(
+        'self',
+        on_delete=models.CASCADE,
+        related_name='duplicates',
+        null=True,
+        blank=True,
+        default=None,
+        verbose_name=_('duplicate of the person'),
+        help_text=_('duplicate of the record of the person'),
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:148]: added

File: person/models.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        if cleaned_data.get('is_main') and cleaned_data.get('duplicate_of'):
+            raise ValidationError('Record of Person cannot be main and duplicate of')
+
+    def save(self, *args, **kwargs):
+        self.full_name = f'{self.last_name} {self.first_name} {self.middle_name}'
+        self.full_name_original = f'{self.last_name_original} {self.first_name_original} {self.middle_name_original}'
+        super().save(*args, **kwargs)
+
+    class Meta:
+        unique_together = ('full_name', 'date_of_birth', 'taxpayer_number')
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:165]: added verbose_name_plural

File: person/models.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+        help_text=_('gender of the person')
+    )
+    is_dead = models.BooleanField(
+        _('is dead'),
+        default=False,
+        help_text=_('is person dead')
+    )
+    citizenship_status = models.PositiveSmallIntegerField(
+        _('citizenship status of the person'),
+        choices=CitizenshipStatus.choices,
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:106]: added

File: person/models.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+        _('first name'),
+        max_length=50,
+        validators=[RegexValidator(regex=r'[A-Za-z]')],
+        help_text='first name of the person',
+    )
+    middle_name = models.CharField(
+        _('middle name'),
+        max_length=50,
+        blank=True,
+        validators=[RegexValidator(regex=r'[A-Za-z]')],
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:25]: changed

File: person/models.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+        help_text=_('duplicate of the record of the person'),
+    )
+
+    def __str__(self):
+        return f'{self.full_name} from {self.residence}'
+
+    def clean(self):
+        cleaned_data = super().clean()
+        if cleaned_data.get('is_main') and cleaned_data.get('duplicate_of'):
+            raise ValidationError('Record of Person cannot be main and duplicate of')
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:157]: changed

[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:157]: thank you, fixed

File: person/models.py [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+
+    def save(self, *args, **kwargs):
+        self.full_name = f'{self.last_name} {self.first_name} {self.middle_name}'
+        self.full_name_original = f'{self.last_name_original} {self.first_name_original} {self.middle_name_original}'
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/models.py:161]: thank you, fixed

File: person/models.py [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+    @property
+    def date_of_birth_str(self):
+        DOBAccuracyChoices.date_of_birth_str(self.date_of_birth, self.date_of_birth_accuracy)
+
+    def __str__(self):
+        return f'[{self.id}] {self.full_name}'
+
+    def clean(self):
+        if self.is_main and self.duplicate_of:
+            raise ValidationError({'duplicate of': 'Record cannot be main and duplicate'})
```
Comment (roman-tiukh) [inline person/models.py:159]:
```suggestion
            raise ValidationError({'duplicate_of': 'Record cannot be main and duplicate'})
```

**Review by roman-tiukh — APPROVED** [review]

File: data_ocean/models.py [hunk 17]
```diff
@@ -44,6 +45,11 @@ def truncate_cascade(cls):
         with connection.cursor() as c:
             c.execute('TRUNCATE TABLE "{0}" CASCADE'.format(cls._meta.db_table))
 
+    def strip_text_fields(self):
+        for field in self._meta.fields:
+            if field is str:
```
Comment (roman-tiukh) [inline data_ocean/models.py:50]:
```suggestion
            if isinstance(field, CharField):
```

File: person/models.py [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+        null=True,
+        blank=True,
+        default=None,
+        verbose_name='duplicate of the persons',
+        help_text='duplicate of the record of the persons',
+    )
+
+    @property
+    def date_of_birth_str(self):
+        DOBAccuracyChoices.date_of_birth_str(self.date_of_birth, self.date_of_birth_accuracy)
```
Comment (roman-tiukh) [inline person/models.py:152]:
```suggestion
        return DOBAccuracyChoices.date_of_birth_str(self.date_of_birth, self.date_of_birth_accuracy)
```

**Review by EmmaSamoilova — APPROVED** [review]

File: person/models.py (hunk 18 above)
Comment (IegorSoboliev) [reply person/models.py:152]: thank you)

File: data_ocean/models.py (hunk 17 above)
Comment (IegorSoboliev) [reply data_ocean/models.py]: that was only direction, not implementation))

File: business_register/locale/uk/LC_MESSAGES/django.po [hunk 19]
```diff
@@ -374,7 +374,7 @@ msgid "is pep"
 msgstr "є публічним діячем"
 
 #: business_register/models/pep_models.py:92
-msgid "associated persons"
+msgid "associated person"
```
Comment (roman-tiukh) [inline business_register/locale/uk/LC_MESSAGES/django.po:377]: persons

**Pull request merged**